subsys/bluetooth/services/hids_c.c

```diff
@@ -0,0 +1,864 @@
+			LOG_DBG("Creating report at index %u, handle: 0x%.4x",
+				rep_cnt, gatt_chrc->handle);
+			ret = rep_new(hids_c,
+				      &(hids_c->rep_info[rep_cnt]),
+				      dm,
+				      gatt_chrc);
+			if (ret) {
+				LOG_ERR("Cannot create report, error: %d", ret);
+				return ret;
+			}
+			++rep_cnt;
+		}
+	}
+
+	LOG_DBG("Report memory allocated, entities used: %u",
+		k_mem_slab_num_used_get(&bt_gatt_hids_c_reports_mem));
+
+	/* Finally - save connection object */
+	hids_c->conn = bt_gatt_dm_conn_get(dm);
+
+	/** @todo This is temporary, as normally there retrieving
+	 * additional data from reports shall start
+	 */
+	hids_c->ready = true;
+	hids_c->ready_cb(hids_c);
+	return 0;
+}
+
+void bt_gatt_hids_c_init(struct bt_gatt_hids_c *hids_c,
+			const struct bt_gatt_hids_c_init_params *params)
+{
+	memset(hids_c, 0, sizeof(*hids_c));
+	hids_c->ready_cb      = params->ready_cb;
+	hids_c->prep_error_cb = params->prep_error_cb;
+	hids_c->pm_update_cb  = params->pm_update_cb;
+	k_sem_init(&hids_c->read_params_sem, 1, 1);
+}
+
+int bt_gatt_hids_c_handles_assign(struct bt_gatt_dm *dm,
+				  struct bt_gatt_hids_c *hids_c)
+{
+	int ret = handles_assign_internal(dm, hids_c);
+
+	if (ret) {
+		bt_gatt_hids_c_release(hids_c);
+	}
+	return ret;
+}
+
+void bt_gatt_hids_c_release(struct bt_gatt_hids_c *hids_c)
+{
+	if (hids_c->rep_info) {
+		u8_t rep_cnt = hids_c->rep_cnt;
+
+		while (rep_cnt) {
+			rep_free(&(hids_c->rep_info[--rep_cnt]));
+		}
+
+		hids_c->rep_cnt = 0;
+		repptr_array_free(hids_c->rep_info);
+		hids_c->rep_info = NULL;
+	}
+	if (hids_c->rep_boot.kbd_inp) {
+		rep_free(&(hids_c->rep_boot.kbd_inp));
+	}
+	if (hids_c->rep_boot.kbd_out) {
+		rep_free(&(hids_c->rep_boot.kbd_out));
+	}
+	if (hids_c->rep_boot.mouse_inp) {
+		rep_free(&(hids_c->rep_boot.mouse_inp));
+	}
+	LOG_DBG("Report memory released, entities used: %u",
+		k_mem_slab_num_used_get(&bt_gatt_hids_c_reports_mem));
+
+	memset(&hids_c->info_val, 0, sizeof(hids_c->info_val));
+	memset(&hids_c->handlers, 0, sizeof(hids_c->handlers));
+	hids_c->map_cb  = NULL;
+	hids_c->ready   = false;
+	hids_c->conn    = NULL;
+	k_sem_give(&hids_c->read_params_sem);
+}
+
+void bt_gatt_hids_c_abort_all(struct bt_gatt_hids_c *hids_c)
+{
+
+}
+
+bool bt_gatt_hids_c_assign_check(const struct bt_gatt_hids_c *hids_c)
+{
+	return hids_c->conn != NULL;
+}
+
+bool bt_gatt_hids_ready_check(const struct bt_gatt_hids_c *hids_c)
+{
+	return hids_c->ready;
+}
+
+/**
+ * @brief Process report read
+ *
+ * Internal function to process report read and pass it further.
+ *
+ * @param conn   Connection handler.
+ * @param err    Read ATT error code.
+ * @param params Notification parameters structure - the pointer
+ *               to the structure provided to read function.
+ * @param data   Pointer to the data buffer.
+ * @param length The size of the received data.
+ *
+ * @retval BT_GATT_ITER_STOP     Stop notification
+ * @retval BT_GATT_ITER_CONTINUE Continue notification
```

Comment:
Not returned by this function.

Reply:
This is callback function and the documentation is how the values would be interpreted by the caller.